py-polars/polars/selectors.py

```diff
@@ -595,6 +596,109 @@ def by_dtype(
+    │ key ┆ c00 ┆ c01 ┆ c02 ┆ … ┆ c96  ┆ c97  ┆ c98  ┆ c99  │
+    │ --- ┆ --- ┆ --- ┆ --- ┆   ┆ ---  ┆ ---  ┆ ---  ┆ ---  │
+    │ str ┆ f64 ┆ f64 ┆ f64 ┆   ┆ f64  ┆ f64  ┆ f64  ┆ f64  │
+    ╞═════╪═════╪═════╪═════╪═══╪══════╪══════╪══════╪══════╡
+    │ abc ┆ 0.0 ┆ 0.5 ┆ 1.0 ┆ … ┆ 48.0 ┆ 48.5 ┆ 49.0 ┆ 49.5 │
+    └─────┴─────┴─────┴─────┴───┴──────┴──────┴──────┴──────┘
+
+    Select columns by index ("key" column and the two first/last columns):
+
+    >>> df.select(cs.by_index(0, 1, 2, -2, -1))
+    shape: (1, 5)
+    ┌─────┬─────┬─────┬──────┬──────┐
+    │ key ┆ c00 ┆ c01 ┆ c98  ┆ c99  │
+    │ --- ┆ --- ┆ --- ┆ ---  ┆ ---  │
+    │ str ┆ f64 ┆ f64 ┆ f64  ┆ f64  │
+    ╞═════╪═════╪═════╪══════╪══════╡
+    │ abc ┆ 0.0 ┆ 0.5 ┆ 49.0 ┆ 49.5 │
+    └─────┴─────┴─────┴──────┴──────┘
+
+    Select the "key" column and use a `range` object to select various columns.
+    Note that you can freely mix and match integer indices and `range` objects:
+
+    >>> df.select(cs.by_index(0, range(1, 101, 20)))
+    shape: (1, 6)
+    ┌─────┬─────┬──────┬──────┬──────┬──────┐
+    │ key ┆ c00 ┆ c20  ┆ c40  ┆ c60  ┆ c80  │
+    │ --- ┆ --- ┆ ---  ┆ ---  ┆ ---  ┆ ---  │
+    │ str ┆ f64 ┆ f64  ┆ f64  ┆ f64  ┆ f64  │
+    ╞═════╪═════╪══════╪══════╪══════╪══════╡
+    │ abc ┆ 0.0 ┆ 10.0 ┆ 20.0 ┆ 30.0 ┆ 40.0 │
+    └─────┴─────┴──────┴──────┴──────┴──────┘
+
+    >>> df.select(cs.by_index(0, range(101, 0, -25)))
+    shape: (1, 5)
+    ┌─────┬──────┬──────┬──────┬─────┐
+    │ key ┆ c75  ┆ c50  ┆ c25  ┆ c00 │
+    │ --- ┆ ---  ┆ ---  ┆ ---  ┆ --- │
+    │ str ┆ f64  ┆ f64  ┆ f64  ┆ f64 │
+    ╞═════╪══════╪══════╪══════╪═════╡
+    │ abc ┆ 37.5 ┆ 25.0 ┆ 12.5 ┆ 0.0 │
+    └─────┴──────┴──────┴──────┴─────┘
+
+    Select all columns *except* for the even-indexed ones:
+
+    >>> df.select(~cs.by_index(range(1, 100, 2)))
+    shape: (1, 51)
+    ┌─────┬─────┬─────┬─────┬───┬──────┬──────┬──────┬──────┐
+    │ key ┆ c01 ┆ c03 ┆ c05 ┆ … ┆ c93  ┆ c95  ┆ c97  ┆ c99  │
+    │ --- ┆ --- ┆ --- ┆ --- ┆   ┆ ---  ┆ ---  ┆ ---  ┆ ---  │
+    │ str ┆ f64 ┆ f64 ┆ f64 ┆   ┆ f64  ┆ f64  ┆ f64  ┆ f64  │
+    ╞═════╪═════╪═════╪═════╪═══╪══════╪══════╪══════╪══════╡
+    │ abc ┆ 0.5 ┆ 1.5 ┆ 2.5 ┆ … ┆ 46.5 ┆ 47.5 ┆ 48.5 ┆ 49.5 │
+    └─────┴─────┴─────┴─────┴───┴──────┴──────┴──────┴──────┘
+    """
+    all_indices = []
+    for idx in indices:
+        if isinstance(idx, int):
+            all_indices.append(idx)
+        elif isinstance(idx, range):
+            all_indices.extend(idx)
+        elif isinstance(idx, Collection):
+            for i in idx:
+                if isinstance(i, int):
+                    all_indices.append(i)
+                else:
+                    msg = f"invalid index: {i!r}"
+                    raise TypeError(msg)
+        else:
+            msg = f"invalid index: {idx!r}"
+            raise TypeError(msg)
+
+    return _selector_proxy_(
+        F.nth(all_indices), name="by_index", parameters={"*indices": indices}
```

Comment:
Shouldn't this use `index_cols` instead of `nth`? `all_indices` is a list.

```suggestion
        F.index_cols(all_indices), name="by_index", parameters={"*indices": indices}
```

Reply:
`index_cols` is what `nth` uses internally, but it's not exposed directly on the python API (so there is no `F.index_cols` ;)

Reply:
Ahh yes it calls the Python `nth` 👍